File: android/sdk/src/main/java/com/noctuagames/sdk/Noctua.kt
package com.noctuagames.sdk

import android.app.Activity
import android.content.Context
import android.util.Log
import com.noctuagames.sdk.models.Account
import com.noctuagames.sdk.models.BillingErrorCode
import com.noctuagames.sdk.models.ConsumableType
import com.noctuagames.sdk.models.NoctuaBillingConfig
import com.noctuagames.sdk.models.NoctuaProductDetails
import com.noctuagames.sdk.models.ProductType
import com.noctuagames.sdk.models.NoctuaProductPurchaseStatus
import com.noctuagames.sdk.models.NoctuaPurchaseResult
import com.noctuagames.sdk.presenter.NoctuaPresenter

/**
 * Public API entry point for the Noctua Native SDK.
 *
 * This object exposes a stable surface for Unity and host applications.
 * All business logic is delegated to [NoctuaPresenter].
 *
 */
object Noctua {

    private const val TAG = "NoctuaAPI"
    private lateinit var presenter: NoctuaPresenter

    // ------------------------------------
    // Initialization
    // ------------------------------------

    /**
     * Initializes internal dependency injection (Koin).
     *
     * Must be called before [init].
     */
    fun initApp() {
        presenter.initKoin()
    }

    /**
     * Initializes the Noctua SDK.
     *
     * @param context Application or Activity context.
     * @param publishedApps Optional list of published application identifiers.
     * @param billingConfig Optional billing configuration. If not provided, defaults are used.
     */
    fun init(
        context: Context,
        publishedApps: List<String> = emptyList(),
        billingConfig: NoctuaBillingConfig = NoctuaBillingConfig()
    ) {
        presenter = NoctuaPresenter(context, publishedApps, billingConfig)
        askNotificationPermission(context as Activity)
    }

    // ------------------------------------
    // Lifecycle
    // ------------------------------------

    /** Should be called from the host application's onResume(). */
    fun onResume() = ensureInit { presenter.onResume() }

    /** Should be called from the host application's onPause(). */
    fun onPause() = ensureInit { presenter.onPause() }

    /** Should be called from the host application's onDestroy(). */
    fun onDestroy() = ensureInit { presenter.onDestroy() }

    /** Notifies SDK that device is online. */
    fun onOnline() = ensureInit { presenter.onOnline() }

    /** Notifies SDK that device is offline. */
    fun onOffline() = ensureInit { presenter.onOffline() }

    // ------------------------------------
    // Tracking
    // ------------------------------------

    /**
     * Tracks ad revenue event.
     *
     * @param source Mediation or network source.
     * @param revenue Revenue amount.
     * @param currency ISO currency code (e.g., USD).
     * @param extraPayload Optional additional parameters.
     */
    fun trackAdRevenue(
        source: String,
        revenue: Double,
        currency: String,
        extraPayload: MutableMap<String, Any> = mutableMapOf()
    ) = ensureInit {
        presenter.trackAdRevenue(source, revenue, currency, extraPayload)
    }

    /**
     * Tracks in-app purchase event.
     *
     * @param orderId Unique purchase identifier.
     * @param amount Purchase amount.
     * @param currency ISO currency code.
     * @param extraPayload Optional additional parameters.
     */
    fun trackPurchase(
        orderId: String,
        amount: Double,
        currency: String,
        extraPayload: MutableMap<String, Any> = mutableMapOf()
    ) = ensureInit {
        presenter.trackPurchase(orderId, amount, currency, extraPayload)
    }

    /**
     * Tracks a custom analytics event.
     *
     * @param eventName Event name.
     * @param payload Optional event parameters.
     */
    fun trackCustomEvent(
        eventName: String,
        payload: MutableMap<String, Any> = mutableMapOf()
    ) = ensureInit {
        presenter.trackCustomEvent(eventName, payload)
    }

    /**
     * Tracks a custom event with revenue value.
     * @param eventName Event name.
     * @param revenue Revenue amount.
     * @param currency ISO currency code.
     * @param payload Optional event parameters.
     */
    fun trackCustomEventWithRevenue(
        eventName: String,
        revenue: Double,
        currency: String,
        payload: MutableMap<String, Any> = mutableMapOf()
    ) = ensureInit {
        presenter.trackCustomEventWithRevenue(eventName, revenue, currency, payload)
    }

    // ------------------------------------
    // Accounts
    // ------------------------------------

    /** Returns all stored accounts. */
    fun getAccounts(): List<Account> =
        ifInitialized { presenter.getAccounts() } ?: emptyList()

    /**
     * Returns a specific account.
     *
     * @param userId User identifier.
     * @param gameId Game identifier.
     */
    fun getAccount(userId: Long, gameId: Long): Account? =
        ifInitialized { presenter.getAccount(userId, gameId) }

    /** Returns accounts filtered by user ID. */
    fun getAccountsByUserId(userId: Long): List<Account> =
        ifInitialized { presenter.getAccountsByUserId(userId) } ?: emptyList()

    /** Saves or updates an account. */
    fun putAccount(account: Account) =
        ensureInit { presenter.putAccount(account) }

    /** Deletes an account. */
    fun deleteAccount(account: Account): Int =
        ifInitialized { presenter.deleteAccount(account) } ?: 0

    // ------------------------------------
    // Firebase IDs
    // ------------------------------------

    /** Retrieves Firebase Installation ID. */
    fun getFirebaseInstallationID(onResult: (String) -> Unit) =
        ensureInit { presenter.getFirebaseInstallationID(onResult) }

    /** Retrieves Firebase Analytics Session ID. */
    fun getFirebaseAnalyticsSessionID(onResult: (String) -> Unit) =
        ensureInit { presenter.getFirebaseAnalyticsSessionID(onResult) }

    // ------------------------------------
    // Remote Config
    // ------------------------------------

    /** Returns Firebase Remote Config string value. */
    fun getFirebaseRemoteConfigString(key: String, onResult: (String) -> Unit) =
        ensureInit {
            onResult(presenter.getFirebaseRemoteConfigString(key) ?: "")
        }

    /** Returns Firebase Remote Config boolean value. */
    fun getFirebaseRemoteConfigBoolean(key: String, onResult: (Boolean) -> Unit) =
        ensureInit {
            onResult(presenter.getFirebaseRemoteConfigBoolean(key) ?: false)
        }

    /** Returns Firebase Remote Config double value. */
    fun getFirebaseRemoteConfigDouble(key: String, onResult: (Double) -> Unit) =
        ensureInit {
            onResult(presenter.getFirebaseRemoteConfigDouble(key) ?: 0.0)
        }

    /** Returns Firebase Remote Config long value. */
    fun getFirebaseRemoteConfigLong(key: String, onResult: (Long) -> Unit) =
        ensureInit {
            onResult(presenter.getFirebaseRemoteConfigLong(key) ?: 0L)
        }

    // ------------------------------------
    // Attribution
    // ------------------------------------

    /** Returns Adjust attribution data as JSON string. */
    fun getAdjustAttribution(onResult: (String) -> Unit) =
        ensureInit { presenter.getAdjustAttribution(onResult) }

    // ------------------------------------
    // Notification Permission
    // ------------------------------------

    /**
     * Requests Android 13+ notification permission if required.
     */
    private fun askNotificationPermission(activity: Activity) =
        ensureInit { presenter.askNotificationPermission(activity) }

    // ------------------------------------
    // Internal Tracker Configuration
    // ------------------------------------

    /**
     * Sets additional session parameters.
     *
     * These parameters will be attached to subsequent analytics events.
     *
     * @param extraParams Key-value map containing additional session metadata.
     */
    fun setSessionExtraParams(extraParams: Map<String, Any>) =
        ensureInit { presenter.setSessionExtraParams(extraParams) }

    /**
     * Sets a session tag used for segmentation or internal tracking.
     *
     * @param tag Custom session identifier.
     */
    fun setSessionTag(tag: String) =
        ensureInit { presenter.setSessionTag(tag) }

    /**
     * Returns the currently active session tag.
     *
     * @return Session tag or empty string if SDK is not initialized.
     */
    fun getSessionTag(): String =
        ifInitialized { presenter.getSessionTag() } ?: ""

    /**
     * Sets experiment identifier for A/B testing purposes.
     *
     * @param experiment Experiment name or variant identifier.
     */
    fun setExperiment(experiment: String) =
        ensureInit { presenter.setExperiment(experiment) }

    /**
     * Returns the active experiment identifier.
     *
     * @return Experiment name or empty string if not initialized.
     */
    fun getExperiment(): String =
        ifInitialized { presenter.getExperiment() } ?: ""

    /**
     * Sets a general experiment value.
     *
     * This allows storing multiple experiment values internally.
     *
     * @param experiment Experiment value.
     */
    fun setGeneralExperiment(experiment: String) =
        ensureInit { presenter.setGeneralExperiment(experiment) }

    /**
     * Returns a general experiment value by key.
     *
     * @param experimentKey Key of the experiment.
     * @return Experiment value or empty string if not found or not initialized.
     */
    fun getGeneralExperiment(experimentKey: String): String =
        ifInitialized { presenter.getGeneralExperiment(experimentKey) } ?: ""

    // ------------------------------------
    // Events Local Storage
    // ------------------------------------

    /**
     * Saves externally generated events into SDK local storage.
     *
     * @param jsonString JSON string containing event data.
     */
    fun saveEvents(jsonString: String) =
        ensureInit { presenter.saveEvents(jsonString) }

    /**
     * Retrieves locally stored events.
     *
     * @param onResult Callback returning list of JSON event strings.
     */
    fun getEvents(onResult: (List<String>) -> Unit) =
        ensureInit { presenter.getEvents(onResult) }

    /**
     * Deletes all locally stored events.
     */
    fun deleteEvents() =
        ensureInit { presenter.deleteEvents() }

    // ------------------------------------
    // Events Per-Row Storage (Unlimited)
    // ------------------------------------

    /**
     * Inserts a single event JSON into per-row database storage.
     *
     * @param eventJson Serialized JSON string for one event.
     */
    fun insertEvent(eventJson: String) =
        ensureInit { presenter.insertEvent(eventJson) }

    /**
     * Retrieves a batch of events from database storage.
     *
     * @param limit Maximum number of events to return.
     * @param offset Number of events to skip.
     * @param onResult Callback returning JSON array of event objects.
     */
    fun getEventsBatch(limit: Int, offset: Int, onResult: (String) -> Unit) =
        ensureInit { presenter.getEventsBatch(limit, offset, onResult) }

    /**
     * Deletes specific events by their database IDs.
     *
     * @param idsJson JSON array string of IDs to delete, e.g. "[1,2,3]".
     * @param onResult Callback returning the number of deleted rows.
     */
    fun deleteEventsByIds(idsJson: String, onResult: (Int) -> Unit) =
        ensureInit { presenter.deleteEventsByIds(idsJson, onResult) }

    /**
     * Returns the total count of stored events.
     *
     * @param onResult Callback returning the event count.
     */
    fun getEventCount(onResult: (Int) -> Unit) =
        ensureInit { presenter.getEventCount(onResult) }

    // ------------------------------------
    // In-App Purchases (Billing)
    // ------------------------------------

    /**
     * Initializes the billing service for in-app purchases.
     *
     * @param onPurchaseCompleted Callback when a purchase is completed.
     * @param onPurchaseUpdated Callback when a purchase is updated.
     * @param onProductDetailsLoaded Callback when product details are loaded.
     * @param onQueryPurchasesCompleted Callback when purchases are queried.
     * @param onRestorePurchasesCompleted Callback when restore purchases completes.
     * @param onProductPurchaseStatusResult Callback with the purchase status of a product.
     * @param onServerVerificationRequired Callback when a purchase requires server verification.
     *   The integrator must verify the purchase on their server and then call
     *   [completePurchaseProcessing] with the result.
     * @param onBillingError Callback when a billing error occurs.
     */
    fun initializeBilling(
        onPurchaseCompleted: ((NoctuaPurchaseResult) -> Unit)? = null,
        onPurchaseUpdated: ((NoctuaPurchaseResult) -> Unit)? = null,
        onProductDetailsLoaded: ((List<NoctuaProductDetails>) -> Unit)? = null,
        onQueryPurchasesCompleted: ((List<NoctuaPurchaseResult>) -> Unit)? = null,
        onRestorePurchasesCompleted: ((List<NoctuaPurchaseResult>) -> Unit)? = null,
        onProductPurchaseStatusResult: ((NoctuaProductPurchaseStatus) -> Unit)? = null,
        onServerVerificationRequired: ((NoctuaPurchaseResult, ConsumableType) -> Unit)? = null,
        onBillingError: ((BillingErrorCode, String) -> Unit)? = null
    ) = ensureInit {
        presenter.initializeBilling(onPurchaseCompleted, onPurchaseUpdated, onProductDetailsLoaded, onQueryPurchasesCompleted, onRestorePurchasesCompleted, onProductPurchaseStatusResult, onServerVerificationRequired, onBillingError)
    }

    /**
     * Registers a product with its consumable type.
     *
     * @param productId The product ID from Google Play Console.
     * @param consumableType The type of product (CONSUMABLE, NON_CONSUMABLE, SUBSCRIPTION).
     */
    fun registerProduct(productId: String, consumableType: ConsumableType) =
        ensureInit { presenter.registerProduct(productId, consumableType) }

    /**
     * Queries product details from Google Play.
     *
     * @param productIds List of product IDs to query.
     * @param productType Type of products (INAPP or SUBS).
     */
    fun queryProductDetails(productIds: List<String>, productType: ProductType = ProductType.INAPP) =
        ensureInit { presenter.queryProductDetails(productIds, productType) }

    /**
     * Launches the billing flow for a purchase.
     *
     * @param activity The activity to launch the billing flow from.
     * @param productDetails The product details to purchase.
     */
    fun launchBillingFlow(activity: Activity, productDetails: NoctuaProductDetails) =
        ensureInit { presenter.launchBillingFlow(activity, productDetails) }

    /**
     * Queries existing purchases.
     *
     * @param productType Type of products to query.
     */
    fun queryPurchases(productType: ProductType = ProductType.INAPP) =
        ensureInit { presenter.queryPurchases(productType) }

    /**
     * Acknowledges a purchase.
     *
     * @param purchaseToken The purchase token to acknowledge.
     * @param callback Callback with success status.
     */
    fun acknowledgePurchase(purchaseToken: String, callback: ((Boolean) -> Unit)? = null) =
        ensureInit { presenter.acknowledgePurchase(purchaseToken, callback) }

    /**
     * Consumes a purchase (for consumable products).
     *
     * @param purchaseToken The purchase token to consume.
     * @param callback Callback with success status.
     */
    fun consumePurchase(purchaseToken: String, callback: ((Boolean) -> Unit)? = null) =
        ensureInit { presenter.consumePurchase(purchaseToken, callback) }

    /**
     * Completes purchase processing after server verification.
     *
     * Call this after your server has verified (and acknowledged) the purchase
     * received via [onServerVerificationRequired].
     * - For **consumables**: consumes the purchase client-side so it can be bought again.
     * - For **non-consumables / subscriptions**: no client-side action needed since
     *   the server already acknowledged via the Google Play Developer API.
     *
     * @param purchaseToken The purchase token that was verified.
     * @param consumableType The consumable type of the product.
     * @param verified Whether the server verification succeeded.
     * @param callback Optional callback with success status.
     */
    fun completePurchaseProcessing(
        purchaseToken: String,
        consumableType: ConsumableType,
        verified: Boolean,
        callback: ((Boolean) -> Unit)? = null
    ) = ensureInit {
        presenter.completePurchaseProcessing(purchaseToken, consumableType, verified, callback)
    }

    /**
     * Restores all purchases by querying both INAPP and SUBS purchases.
     * Unacknowledged purchases are automatically processed (acknowledged or consumed)
     * based on their registered consumable type.
     */
    fun restorePurchases() =
        ensureInit { presenter.restorePurchases() }

    /**
     * Gets the purchase status of a specific product.
     * The result is returned via the [onProductPurchaseStatusResult] callback
     * registered in [initializeBilling].
     *
     * @param productId The product ID to check.
     */
    fun getProductPurchaseStatus(productId: String) =
        ensureInit { presenter.getProductPurchaseStatus(productId) }

    /**
     * Reconnects the billing client.
     */
    fun reconnectBilling() =
        ensureInit { presenter.reconnectBilling() }

    /**
     * Disposes the billing service.
     */
    fun disposeBilling() =
        ensureInit { presenter.disposeBilling() }

    /**
     * Checks if billing client is ready.
     *
     * @return True if billing is ready, false otherwise.
     */
    fun isBillingReady(): Boolean =
        ifInitialized { presenter.isBillingReady() } ?: false

    // ------------------------------------
    // Internal Helpers
    // ------------------------------------

    /**
     * Ensures SDK is initialized before executing logic.
     */
    private fun ensureInit(block: () -> Unit) {
        if (!::presenter.isInitialized) {
            Log.e(TAG, "Noctua is not initialized. Call init() first.")
            return
        }
        block()
    }

    /**
     * Executes block only if SDK is initialized.
     */
    private fun <T> ifInitialized(block: () -> T): T? {
        if (!::presenter.isInitialized) {
            Log.e(TAG, "Noctua is not initialized. Call init() first.")
            return null
        }
        return block()
    }
}

File: ios/NoctuaSDK/Sources/Noctua.swift
import Foundation

@objc public class Noctua: NSObject {
    @objc public static func initNoctua() throws {
        if tracker == nil && storeKit == nil && account == nil && session == nil {
            let config = try loadConfig()
            let logger = IOSLogger(category: "Noctua")

            let services = buildServices(config: config, logger: logger)

            tracker = TrackerPresenter(
                config: config,
                trackers: services.trackers,
                noctuaInternal: services.noctuaInternal,
                logger: logger
            )

            storeKit = StoreKitPresenter(
                storeKitService: services.storeKitService,
                logger: logger
            )

            account = AccountPresenter(
                accountRepo: services.accountRepo,
                logger: logger
            )

            session = SessionPresenter(
                config: config,
                adjustSpecific: services.adjustSpecific,
                firebaseQuery: services.firebaseQuery,
                noctuaInternal: services.noctuaInternal,
                logger: logger
            )
        }
    }

    // MARK: - Tracking

    @objc public static func trackAdRevenue(source: String, revenue: Double, currency: String, extraPayload: [String: Any] = [:]) {
        tracker?.trackAdRevenue(source: source, revenue: revenue, currency: currency, extraPayload: extraPayload)
    }

    @objc public static func trackPurchase(orderId: String, amount: Double, currency: String, extraPayload: [String: Any] = [:]) {
        tracker?.trackPurchase(orderId: orderId, amount: amount, currency: currency, extraPayload: extraPayload)
    }

    @objc public static func trackCustomEvent(_ eventName: String, payload: [String: Any] = [:]) {
        tracker?.trackCustomEvent(eventName, payload: payload)
    }

    @objc public static func trackCustomEventWithRevenue(_ eventName: String, revenue: Double, currency: String, payload: [String: Any] = [:]) {
        tracker?.trackCustomEventWithRevenue(eventName, revenue: revenue, currency: currency, payload: payload)
    }

    // MARK: - StoreKit / In-App Purchases

    @objc public static func initializeStoreKit(
        onPurchaseCompleted: ((NoctuaPurchaseResult) -> Void)? = nil,
        onPurchaseUpdated: ((NoctuaPurchaseResult) -> Void)? = nil,
        onProductDetailsLoaded: (([NoctuaProductDetails]) -> Void)? = nil,
        onQueryPurchasesCompleted: (([NoctuaPurchaseResult]) -> Void)? = nil,
        onRestorePurchasesCompleted: (([NoctuaPurchaseResult]) -> Void)? = nil,
        onProductPurchaseStatusResult: ((NoctuaProductPurchaseStatus) -> Void)? = nil,
        onServerVerificationRequired: ((NoctuaPurchaseResult, ConsumableType) -> Void)? = nil,
        onStoreKitError: ((StoreKitErrorCode, String) -> Void)? = nil
    ) {
        storeKit?.initializeStoreKit(
            onPurchaseCompleted: onPurchaseCompleted,
            onPurchaseUpdated: onPurchaseUpdated,
            onProductDetailsLoaded: onProductDetailsLoaded,
            onQueryPurchasesCompleted: onQueryPurchasesCompleted,
            onRestorePurchasesCompleted: onRestorePurchasesCompleted,
            onProductPurchaseStatusResult: onProductPurchaseStatusResult,
            onServerVerificationRequired: onServerVerificationRequired,
            onStoreKitError: onStoreKitError
        )
    }

    @objc public static func registerProduct(productId: String, consumableType: ConsumableType) {
        storeKit?.registerProduct(productId: productId, consumableType: consumableType)
    }

    @objc public static func queryProductDetails(productIds: [String], productType: ProductType = .inapp) {
        storeKit?.queryProductDetails(productIds: productIds, productType: productType)
    }

    @objc public static func purchase(productId: String) {
        storeKit?.purchase(productId: productId)
    }

    @objc public static func queryPurchases(productType: ProductType = .inapp) {
        storeKit?.queryPurchases(productType: productType)
    }

    @objc public static func restorePurchases() {
        storeKit?.restorePurchases()
    }

    @objc public static func getProductPurchaseStatus(productId: String) {
        storeKit?.getProductPurchaseStatus(productId: productId)
    }

    @objc public static func completePurchaseProcessing(
        purchaseToken: String,
        consumableType: ConsumableType,
        verified: Bool,
        callback: ((Bool) -> Void)? = nil
    ) {
        storeKit?.completePurchaseProcessing(
            purchaseToken: purchaseToken,
            consumableType: consumableType,
            verified: verified,
            callback: callback
        )
    }

    @objc public static func disposeStoreKit() {
        storeKit?.disposeStoreKit()
    }

    @objc public static func isStoreKitReady() -> Bool {
        return storeKit?.isStoreKitReady() ?? false
    }

    // MARK: - Accounts

    @objc public static func putAccount(gameId: Int64, playerId: Int64, rawData: String) {
        account?.putAccount(gameId: gameId, playerId: playerId, rawData: rawData)
    }

    @objc public static func getAllAccounts() -> [[String: Any]] {
        return account?.getAllAccounts() ?? []
    }

    @objc public static func getSingleAccount(gameId: Int64, playerId: Int64) -> [String: Any]? {
        return account?.getSingleAccount(gameId: gameId, playerId: playerId)
    }

    @objc public static func deleteAccount(gameId: Int64, playerId: Int64) {
        account?.deleteAccount(gameId: gameId, playerId: playerId)
    }

    // MARK: - Session & Lifecycle

    @objc public static func onOnline() {
        session?.onOnline()
    }

    @objc public static func onOffline() {
        session?.onOffline()
    }

    @objc public static func getFirebaseInstallationID(completion: @escaping (String) -> Void) {
        session?.getFirebaseInstallationID(completion: completion)
    }

    @objc public static func getFirebaseSessionID(completion: @escaping (String) -> Void) {
        session?.getFirebaseSessionID(completion: completion)
    }

    @objc public static func getFirebaseRemoteConfigString(key: String) -> String? {
        return session?.getFirebaseRemoteConfigString(key: key)
    }

    @objc public static func getFirebaseRemoteConfigBoolean(key: String) -> Bool {
        return session?.getFirebaseRemoteConfigBoolean(key: key) ?? false
    }

    @objc public static func getFirebaseRemoteConfigDouble(key: String) -> Double {
        return session?.getFirebaseRemoteConfigDouble(key: key) ?? 0.0
    }

    @objc public static func getFirebaseRemoteConfigLong(key: String) -> Int64 {
        return session?.getFirebaseRemoteConfigLong(key: key) ?? 0
    }

    @objc public static func setSessionTag(tag: String) {
        session?.setSessionTag(tag: tag)
    }

    @objc public static func getSessionTags() -> String? {
        return session?.getSessionTag()
    }

    @objc public static func setExperiment(experiment: String) {
        session?.setExperiment(experiment: experiment)
    }

    @objc public static func getExperiment() -> String? {
        return session?.getExperiment()
    }

    @objc public static func setGeneralExperiment(experiment: String) {
        session?.setGeneralExperiment(experiment: experiment)
    }

    @objc public static func getGeneralExperiment(experimentKey: String) -> String? {
        return session?.getGeneralExperiment(experimentKey: experimentKey)
    }

    @objc public static func setSessionExtraParams(payload: [String: Any]) {
        session?.setSessionExtraParams(payload: payload)
    }

    @objc public static func saveEvents(jsonString: String) {
        session?.saveEvents(jsonString: jsonString)
    }

    @objc public static func getEvents(onResult: @escaping ([String]) -> Void) {
        session?.getEvents(onResult: onResult)
    }

    @objc public static func deleteEvents() {
        session?.deleteEvents()
    }

    // MARK: - Per-Row Event Storage (Unlimited)

    @objc public static func insertEvent(eventJson: String) {
        session?.insertEvent(eventJson: eventJson)
    }

    @objc public static func getEventsBatch(limit: Int32, offset: Int32, onResult: @escaping (String) -> Void) {
        session?.getEventsBatch(limit: limit, offset: offset, onResult: onResult)
    }

    @objc public static func deleteEventsByIds(idsJson: String, onResult: @escaping (Int32) -> Void) {
        session?.deleteEventsByIds(idsJson: idsJson, onResult: onResult)
    }

    @objc public static func getEventCount(onResult: @escaping (Int32) -> Void) {
        session?.getEventCount(onResult: onResult)
    }

    @objc public static func getAdjustCurrentAttribution(completion: @escaping ([String: Any]) -> Void) {
        if let session = session {
            session.getAdjustCurrentAttribution(completion: completion)
        } else {
            completion([:])
        }
    }

    // MARK: - Currency Query (Delegated IAP)

    /// Queries the App Store for a product's currency code using SKProductsRequest.
    /// This is a read-only query — no SKPaymentTransactionObserver is added,
    /// so it will NOT conflict with game developers' own IAP implementations.
    @objc public static func getActiveCurrency(_ productId: String, completion: @escaping (Bool, String) -> Void) {
        if currencyQuery == nil {
            currencyQuery = CurrencyQueryService()
        }
        currencyQuery?.getActiveCurrency(productId: productId, completion: completion)
    }

    // MARK: - Private

    private static var tracker: TrackerPresenter?
    private static var storeKit: StoreKitPresenter?
    private static var account: AccountPresenter?
    private static var session: SessionPresenter?
    private static var currencyQuery: CurrencyQueryService?

    private static func buildServices(config: NoctuaConfig, logger: NoctuaLogger) -> (
        trackers: [TrackerServiceProtocol],
        storeKitService: StoreKitServiceProtocol?,
        adjustSpecific: AdjustSpecificProtocol?,
        firebaseQuery: FirebaseQueryServiceProtocol?,
        noctuaInternal: NoctuaInternalServiceProtocol?,
        accountRepo: AccountRepositoryProtocol
    ) {
        // Initialize NoctuaInternal first (Koin must init before other services)
        let noctuaInternal = NoctuaInternalService()
        noctuaInternal.initialize()

        var trackers: [TrackerServiceProtocol] = []
        var adjustSpecific: AdjustSpecificProtocol? = nil
        var firebaseQuery: FirebaseQueryServiceProtocol? = nil

        // StoreKit Service (StoreKit 2, requires iOS 15+)
        let storeKitService: StoreKitServiceProtocol?
        if config.noctua?.iapDisabled == true {
            storeKitService = nil
            logger.info("StoreKit disabled by config (iapDisabled: true)")
        } else {
            let storeKitConfig = NoctuaStoreKitConfig()
            storeKitService = StoreKitService(config: storeKitConfig, logger: logger)
            logger.info("StoreKitService initialized (StoreKit 2)")
        }

        // AdjustService
        if config.adjust == nil {
            logger.warning("config for AdjustService not found")
        } else if config.adjust?.ios == nil {
            logger.warning("config for AdjustService IOS not found")
        } else {
            do {
                let service = try AdjustService(config: (config.adjust?.ios!)!, logger: logger)
                trackers.append(service)
                adjustSpecific = service
                logger.info("AdjustService initialized")
            } catch AdjustServiceError.adjustNotFound {
                logger.warning("Adjust disabled, Adjust module not found")
            } catch AdjustServiceError.invalidConfig(let message) {
                logger.warning("Adjust disabled, invalid Adjust config: \(message)")
            } catch {
                logger.warning("Adjust disabled, unknown error")
            }
        }

        // FirebaseService
        if config.firebase == nil {
            logger.warning("config for FirebaseService not found")
        } else if config.firebase?.ios == nil {
            logger.warning("config for FirebaseService not found")
        } else {
            do {
                let service = try FirebaseService(config: (config.firebase?.ios!)!, logger: logger)
                trackers.append(service)
                firebaseQuery = service
                logger.info("FirebaseService initialized")
            } catch FirebaseServiceError.firebaseNotFound {
                logger.warning("Firebase disabled, Firebase module not found")
            } catch FirebaseServiceError.invalidConfig(let message) {
                logger.warning("Firebase disabled, invalid Firebase config: \(message)")
            } catch {
                logger.warning("Firebase disabled, unknown error")
            }
        }

        // FacebookService
        if config.facebook == nil {
            logger.warning("config for FacebookService not found")
        } else if config.facebook?.ios == nil {
            logger.warning("config for FacebookService not found")
        } else {
            do {
                let service = try FacebookService(config: (config.facebook?.ios!)!, logger: logger)
                trackers.append(service)
                logger.info("FacebookService initialized")
            } catch FacebookServiceError.facebookNotFound {
                logger.warning("Facebook disabled, Facebook module not found")
            } catch FacebookServiceError.invalidConfig(let message) {
                logger.warning("Facebook disabled, invalid Facebook config: \(message)")
            } catch {
                logger.warning("Facebook disabled, unknown error")
            }
        }

        let accountRepo = AccountRepository(logger: logger)

        return (trackers, storeKitService, adjustSpecific, firebaseQuery, noctuaInternal, accountRepo)
    }
}

func loadConfig() throws -> NoctuaConfig {
    let firstPath = Bundle.main.path(forResource: "/Data/Raw/noctuagg", ofType: "json")
    let secondPath = Bundle.main.path(forResource: "noctuagg", ofType: "json")

    guard let path = firstPath ?? secondPath else {
        throw ConfigurationError.fileNotFound
    }

    let config: NoctuaConfig

    do {
        let data = try Data(contentsOf: URL(fileURLWithPath: path))
        config = try JSONDecoder().decode(NoctuaConfig.self, from: data)
    }
    catch DecodingError.valueNotFound(let type, let context) {
        throw ConfigurationError.missingKey("type: \(type), desc: \(context.debugDescription)")
    }
    catch DecodingError.keyNotFound(let key, let context) {
        throw ConfigurationError.missingKey("type: \(key), desc: \(context.debugDescription)")
    }
    catch {
        throw ConfigurationError.invalidFormat
    }

    if config.clientId.isEmpty {
        throw ConfigurationError.missingKey("clientId")
    }

    return config
}
